CodePush.m

```diff
@@ -554,20 +523,65 @@ - (void)savePendingUpdate:(NSString *)packageHash
+            NSString *previousStatusReportIdentifier = [CodePushStatusReport getPreviousStatusReportIdentifier];
+            if (currentPackageIdentifier) {
+                if (previousStatusReportIdentifier == nil) {
+                    [CodePushStatusReport recordDeploymentStatusReported:currentPackageIdentifier];
+                    resolve(@{
+                              @"package": currentPackage,
+                              @"status": DeploymentSucceeded
+                              });
+                    return;
+                } else if (![previousStatusReportIdentifier isEqualToString:currentPackageIdentifier]) {
+                    [CodePushStatusReport recordDeploymentStatusReported:currentPackageIdentifier];
+                    if ([CodePushStatusReport isStatusReportIdentifierCodePushLabel:previousStatusReportIdentifier]) {
+                        NSString *previousDeploymentKey = [CodePushStatusReport getDeploymentKeyFromStatusReportIdentifier:previousStatusReportIdentifier];
+                        NSString *previousLabel = [CodePushStatusReport getVersionLabelFromStatusReportIdentifier:previousStatusReportIdentifier];
+                        resolve(@{
+                                  @"package": currentPackage,
+                                  @"status": DeploymentSucceeded,
+                                  @"previousDeploymentKey": previousDeploymentKey,
+                                  @"previousLabelOrAppVersion": previousLabel
+                                  });
+                    } else {
+                        // Previous status report was with a binary app version.
+                        resolve(@{
+                                  @"package": currentPackage,
+                                  @"status": DeploymentSucceeded,
+                                  @"previousLabelOrAppVersion": previousStatusReportIdentifier
+                                  });
+                    }
+                    return;
+                }
             }
         }
     } else if (isRunningBinaryVersion || [_bridge.bundleURL.scheme hasPrefix:@"http"]) {
         // Check if the current appVersion has been reported.
         NSString *appVersion = [[CodePushConfig current] appVersion];
-        if ([self isDeploymentStatusNotYetReported:appVersion]) {
-            [self recordDeploymentStatusReported:appVersion];
+        NSString *previousStatusReportIdentifier = [CodePushStatusReport getPreviousStatusReportIdentifier];
+        if (previousStatusReportIdentifier == nil) {
+            [CodePushStatusReport recordDeploymentStatusReported:appVersion];
             resolve(@{ @"appVersion": appVersion });
             return;
+        } else if (![previousStatusReportIdentifier isEqualToString:appVersion]) {
+            [CodePushStatusReport recordDeploymentStatusReported:appVersion];
+            if ([CodePushStatusReport isStatusReportIdentifierCodePushLabel:previousStatusReportIdentifier]) {
```

Comment:
If it is an appVersion identifier, it does not include a colon. Hence the name `isStatusReportIdentifierCodePushLabel`, it tells you whether it is a CodePush label or it is just a native app version. I agree it would be much cleaner if all this was in the CodePushStatusReport class, i'll push a change in a bit to move that there.